### ISSUE TITLE
Make callbacks consistent

### ISSUE THREAD

**Issue by darkwebdev** [eBay/ebayui-core-react#171]
## eBayUI Core Marko
For example ebay-menu/ebay-menu-button (parameters passed via an object):
```
onChange (radio): { el, index, checked, originalEvent }
onChange (checkbox): { el, [indexes], [checked], originalEvent }
onChange (not radio or checkbox): { el, index, checked, originalEvent }
onSelect: { el, originalEvent, index }
keydown: { el, index, checked }
```

## eBayUI Core React
Currently (parameters passed via comma-separated arguments):
```
onChange: (index, value, event)
```
I'd suggest to switch from arguments to object like in Marko version and deprecate old variant:
```
onChange: ({ index: number, value: string | number, originalEvent: ChangeEvent })
onChange (checkboxes): ({ indexes: number[], values: string[] | number[], originalEvent: ChangeEvent })
onKeyDown: ({ index: number, key: string, originalEvent: KeyboardEvent })
```
That's not only consistent with marko version, but also makes it easier to introduce new parameters.
For now we can check if there's only one argument which is an object -> treat it the new way, otherwise use deprecated way. So that we don't have to break compatibility yet.

**Comment by aaron-nance** [comment]
I'm definitely for making the API more closely match what Marko has just for the sake of making it simpler to migrate.

**Comment by abiyasa** [comment]
I would against switching from arguments to object. 

While that will make our API similar to Marko but this might cause lots of problem to us:

- I never seen any React UI library which returns 1 argument object for its event handler. From most popular libraries like [MaterialUI](https://mui.com/), [SemanticUI](https://react.semantic-ui.com/), [ChakraUI](https://chakra-ui.com/), [ReachUI](https://reach.tech/), and [Microsoft FluentUI](https://developer.microsoft.com/en-us/fluentui#/controls/web), all of them are using multiple arguments on its event handler.
- Different idiom: We're using React and there are many things how it's different than Marko. Marko does event emitter with `emit()` while in React, we're calling the event handler function directly. The event that emitted by Marko component will be propagated automatically to its parent components, something that React does not have. So while we're always trying to use the same API, there will be some difference due to React vs Marko idiom. 
- We will have more problems with TypeScript. For example, our EbayButton extends some props from React HTMLButtonElement which has `onChange(e: SyntheticMouseEvent)`, or other component `onChange(e: ChangeEvent<HTMLInputElement>)`. If we switch to object, we need to omit and redefine all React HTML element's event handlers on our component.
- We also will loose some flexibility for consistency as we support pass-through attribute on most of our components. For example, EbayButton will pass any remaining props to the output `<a>` or `<button>`. If we want to handle `onTouchStart`, at the moment we just pass that to EbayButton:

  ```jsx
  <EbayButton onTouchMove={(e) => { /* do something with synthetic event */}} />
  ```

  By React standard, all event handler will return event object (see [React type definition for EventHandler](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react/index.d.ts#L1344-L1346)). So if we want to have all event handlers consistent to always return 1 argument object instead of event object, this will be very hard for us.

I would suggest to:

- if our event handler returns the event object, it should be the first one in the event handler argument.
- Keep the rest of arguments as close as possible with Marko event arguments (e.g index, selected, ...)

**Comment by darkwebdev** [comment]
I agree with Abi here, let's make one consistent callback style by defining which arguments and type to use:
```
onClick: (e: React.MouseEvent, ...rest) => void
onKeyDown: (e: React.KeyboardEvent<HTMLButtonElement | HTMLAnchorElement>, ...rest) => void
onChange: (e: ChangeEvent<HTMLInputElement>, ...rest) => void
```

**Comment by darkwebdev** [comment]
Discussion with marko team:
https://ebay-eng.slack.com/archives/C0RKTR82W/p1643655338741799

**Comment by HenriqueLimas** [comment]
Agree with the proposal of having an object as a second argument, this make the argument number consistent around components. But I would suggest to not export handler types with generics. Every component should export their own type and abstract the element that is being used:

```tsx
import { EbayButton, EbayButtonClickEventHandler } from '@ebay/ui-core-react/ebay-button'

const CustomButton = (props : { onClick: EbayButtonClickEventHandler }) => (
   <EbayButton onClick={props.onClick} />
)
```

In that way application doesn't need to know about the element that is being used internally. A button can be a `<button>` but also `<a class="fake-btn">`, we should abstract that out from the application owner.

We can have those base event handlers that we use internally and maybe export them outside, but application should use the event handlers of each component.

**Comment by darkwebdev** [comment]
will be implemented in #210
